internal/core/items/embed.rs

```diff
@@ -0,0 +1,261 @@
+// Copyright © SixtyFPS GmbH <info@slint-ui.com>
+// SPDX-License-Identifier: GPL-3.0-only OR LicenseRef-Slint-commercial
+
+/*!
+This module contains the builtin `Embed` and related items
+
+When adding an item or a property, it needs to be kept in sync with different place.
+Lookup the [`crate::items`] module documentation.
+*/
+use super::{Item, ItemConsts, ItemRc, RenderingResult};
+use crate::component::{ComponentRc, ComponentWeak, IndexRange};
+use crate::component_factory::ComponentFactoryBox;
+use crate::input::{
+    FocusEvent, FocusEventResult, InputEventFilterResult, InputEventResult, KeyEvent,
+    KeyEventResult, MouseEvent,
+};
+use crate::item_rendering::CachedRenderingData;
+use crate::item_tree::{ItemVisitorVTable, TraversalOrder, VisitChildrenResult};
+use crate::layout::{LayoutInfo, Orientation};
+use crate::lengths::{LogicalLength, LogicalPoint, LogicalRect, LogicalSize};
+use crate::properties::PropertyTracker;
+#[cfg(feature = "rtti")]
+use crate::rtti::*;
+use crate::window::WindowAdapter;
+use crate::Property;
+use alloc::rc::Rc;
+use const_field_offset::FieldOffsets;
+use core::cell::RefCell;
+use core::pin::Pin;
+use i_slint_core_macros::*;
+use vtable::VRefMut;
+
+#[repr(C)]
+#[derive(FieldOffsets, Default, SlintElement)]
+#[pin]
+/// The implementation of the `Embed` element
+pub struct Embed {
+    pub component_factory: Property<ComponentFactoryBox>,
+
+    pub x: Property<LogicalLength>,
+    pub y: Property<LogicalLength>,
+    pub width: Property<LogicalLength>,
+    pub height: Property<LogicalLength>,
+
+    pub cached_rendering_data: CachedRenderingData,
+
+    data: EmbedDataBox,
+    component: RefCell<Option<ComponentRc>>,
+}
+
+impl Embed {
+    pub fn ensure_updated(self: Pin<&Self>, parent_component: &ComponentWeak, index: usize) {
+        let is_dirty =
+            self.data.component_tracker.borrow().as_ref().map(|t| t.is_dirty()).unwrap_or(true);
+        if is_dirty {
+            self.update_from_factory(parent_component, index);
+        }
+    }
+
+    fn update_from_factory(self: Pin<&Self>, parent_component: &ComponentWeak, index: usize) {
+        let t = self
+            .data
+            .component_tracker
+            .borrow_mut()
+            .take()
```

Comment:
take?   that sounds quite complicated...  where did you see this pattern in use?